Ensure carousel filters are manifested to lists at each step

peppy opened the pull request:
The original `IEnumerable` flow prioritised slight performance gains, but a filter's implementation could actually make this detrimental to overall performance.

I noticed in passing that there were already potentially multiple enumerations, via `updateYPositions` and the final `ToList` call. Rather than faffing around, let's keep things simple and require lists.

In benchmarking, the difference is (currently) negiligible. Slight improvement if anything.

peppy committed:
Ensure carousel filters are manifested to lists at each step

The original `IEnumerable` flow prioritised slight performance gains,
but a filter's implementation could actually make this detrimental to
overall performance.

I noticed in passing that there were already potentially multiple
enumerations, via `updateYPositions` and the final `ToList` call. Rather
than faffing around, let's keep things simple and require lists.

In benchmarking, the difference is (currently) negiligible. Slight
improvement if anything.

bdach commented on osu.Game/Graphics/Carousel/Carousel.cs:
```diff
@@ -274,7 +274,13 @@ await Task.Run(async () =>
+                        //
+                        // A future improvement may be passing a reference list through each filter rather than copying each time,
+                        // but this is the safest approach.
+                        items = filteredItems as List<CarouselItem> ?? filteredItems.ToList();
```
Any reason to be playing these cast games rather than change `ICarouselFilter.Run()` to return `Task<List<CarouselItem>>`?

peppy replied:
I felt like there was an explicit reason I didn't do this (maybe just the input and output being different feels weird?), but let me give it a shot

peppy committed:
Change `ICarouselFilter` interface return type rather than manual `ToList`

I was hung up on keeping `IEnumerable` but there doesn't seem to be a
good reason to do so.

bdach reviewed (approved):
will wait for CI
